src/dlsnode/storage/util/Promise.d

```diff
@@ -0,0 +1,318 @@
+        this.bytes_last_read += result.length;
+    }
+
+
+    /***************************************************************************
+
+        Method used by the producer to mark the promise as fulfilled,
+        with the error or without.
+
+        Params:
+            error = indicator if the error has happened.
+
+    ***************************************************************************/
+
+    public void fulfilled (bool error)
+    {
+        this.state = error? State.Error : State.PromiseFulfilled;
+    }
+
+
+    /***************************************************************************
+
+        Returns:
+            number of bytes missing for the promise to be fulfilled.
+
+    ***************************************************************************/
+
+    public size_t dataMissing ()
+    {
+        return this.data_buffer.length - this.bytes_last_read;
+    }
+
+
+    /**********************************************************************
+
+        Gets the result from the promise.
+
+        Returns:
+            result of the finished async IO operation
+
+    **********************************************************************/
+
+    private void[] result ()
+    {
+        verify(this.promise_fulfilled());
+        return this.data_buffer[0..this.bytes_last_read];
+    }
+
+
+    /**********************************************************************
+
+        Checks if the producer has finished processing and it's safe
+        to collect the results.
+
+    **********************************************************************/
+
+    private bool promise_fulfilled ()
+    {
+        return this.state == State.PromiseFulfilled || this.state == State.Error;
+    }
+
+
+    /**********************************************************************
+
+        Check if there was any error during processing.
+
+        Returns:
+            true if there was an error during the processing
+
+    ***********************************************************************/
+
+    private bool error ()
+    {
+        verify(this.promise_fulfilled());
+        return this.state == State.Error;
+    }
+
+
+    /**********************************************************************
+
+        Returns:
+            future associated with this promise.
+
+    **********************************************************************/
+
+    public Future!(T) getFuture(T)()
+    {
+        return Future!(T)(this);
+    }
+}
+
+/**************************************************************************
+
+    Struct encapsulating "future" result that promise will provide. It's the
+    other side of the promise/future coin - promise is in the ownership of
+    the producer and future is in the ownership of the consumer.
+
+**************************************************************************/
+
+struct Future(T = void[])
+{
+    /// Promise that owns this future
+    private Promise* promise;
```

Comment:
Is it ever valid for this to be null? (Invariant?)

Reply:
Yeah, it can happen, in case of default initialized future struct used as a class variable.